browser/brave_shields/reduce_language_navigation_throttle.cc

```diff
@@ -0,0 +1,114 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_shields/reduce_language_navigation_throttle.h"
+
+#include <memory>
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/threading/thread_task_runner_handle.h"
+#include "brave/browser/brave_browser_process.h"
+#include "brave/components/brave_shields/browser/brave_farbling_service.h"
+#include "brave/components/brave_shields/browser/brave_shields_util.h"
+#include "brave/components/brave_shields/common/features.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/content_settings/core/browser/host_content_settings_map.h"
+#include "components/language/core/browser/language_prefs.h"
+#include "components/language/core/browser/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "net/http/http_util.h"
+#include "third_party/blink/renderer/platform/language.h"
+#include "third_party/blink/renderer/platform/wtf/text/wtf_string.h"
+
+namespace brave_shields {
+
+// static
+std::unique_ptr<ReduceLanguageNavigationThrottle>
+ReduceLanguageNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* navigation_handle,
+    HostContentSettingsMap* content_settings) {
+  if (!base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveReduceLanguage))
+    return nullptr;
+  return std::make_unique<ReduceLanguageNavigationThrottle>(navigation_handle,
+                                                            content_settings);
+}
+
+ReduceLanguageNavigationThrottle::ReduceLanguageNavigationThrottle(
+    content::NavigationHandle* navigation_handle,
+    HostContentSettingsMap* content_settings)
+    : content::NavigationThrottle(navigation_handle),
+      content_settings_(content_settings) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+ReduceLanguageNavigationThrottle::~ReduceLanguageNavigationThrottle() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+content::NavigationThrottle::ThrottleCheckResult
+ReduceLanguageNavigationThrottle::WillStartRequest() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  UpdateHeaders();
+  return content::NavigationThrottle::PROCEED;
+}
+
+content::NavigationThrottle::ThrottleCheckResult
+ReduceLanguageNavigationThrottle::WillRedirectRequest() {
+  return WillStartRequest();
```

Comment:
I was a bit confused by this at first, I think for clarity we should call UpdateHeaders() in both methods and you can move the DCHECK into `UpdateHeaders`. Otherwise at first glance it looks like we're doing something odd here at least to me

Reply:
changed in https://github.com/brave/brave-core/pull/12234/commits/fc7510cffc3499d49b9c85ac7e3536e4cc0b5d01